# 4.5 Data Consistency Checks

### This script contains the following points:
## Step 1
- **01. Importing Libraries & Step 1**
- **02. Finding Missing Values**
- **03. Duplicates**
## Step 2
- **04. Run the df.describe() function on your df_ords dataframe**
## Step 3
- **05. Check for mixed-type data**
## Step 5
- **06. Run a check for missing values**
## Step 6
- **07. Address the missing values**
## Step 7
- **08. Run a check for duplicate values**
## Step 9
- **09. Export the final cleaned datasets**

# Step 1

## 01. Importing Libraries

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [2]:
# Turn Path into a String & Assign Python Shortcut
path = '/Users/mgottlieb/Documents/Achievement 4'

In [3]:
# Import the products.csv dataset
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [5]:
# Import the orders_wrangled.csv dataset
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [6]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

In [7]:
# Create a test dataframe
df_test = pd.DataFrame()

In [8]:
# Create a mixed typed column
df_test['mix'] = ['a', 'b', 1, True]

In [10]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [12]:
# The UPDATED function for checking whether a dataframe contains any mixed-type columns
for col in df_test.columns.tolist():
    weird = df_test[col].apply(lambda x: type(x) != type(df_test[col].iloc[0])).any()
    if weird:
        print(col)


mix


In [13]:
# Convert column to string
df_test['mix'] = df_test['mix'].astype('str')

In [14]:
# Checking the column type
print(df_test['mix'].dtype)

object


In pandas, the **object** data type is used for columns that contain strings (text) and other data types for which no other specific dtype is designated (like mixed numeric and non-numeric values).

## 02. Finding Missing Values

In [15]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [17]:
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [18]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [19]:
df_prods.shape

(49693, 5)

In [21]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [22]:
df_prods_clean.shape

(49677, 5)

## 03. Duplicates

In [23]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [25]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [26]:
df_prods_clean.shape

(49677, 5)

In [27]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [28]:
df_prods_clean_no_dups.shape

(49672, 5)

In [29]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))


# Step 2

## 04. Run the df.describe() function on your df_ords dataframe

In [30]:
# Df_ords dataframe generated descriptive statistics
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

Based on the provided summary statistics for the dataframe, here are the key points and potential issues to investigate:


**Unnamed: 0**
- This is an index column, it might be redundant. Check to see if this is or isn't


**order_number**
- Verify if the upper limit of 100 is a system constraint or if it's genuinely representative of user activity.


**days_since_prior_order**
- Investigate the meaning of 0 values. Determine if they represent first-time orders or if they indicate missing values or errors. Verify if the maximum value of 30 is expected.


# Step 3

## 05. Check for mixed-type data

In [31]:
# The function for checking whether a dataframe contains any mixed-type columns
for col in df_ords.columns.tolist():
    weird = df_ords[col].apply(lambda x: type(x) != type(df_ords[col].iloc[0])).any()
    if weird:
        print(col)

Unnamed: 0
order_id
user_id
order_number
orders_day_of_week
order_hour_of_day
days_since_prior_order


In [32]:
df_ords.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 8 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   Unnamed: 0              int64  
 1   order_id                int64  
 2   user_id                 int64  
 3   evaluation_status       object 
 4   order_number            int64  
 5   orders_day_of_week      int64  
 6   order_hour_of_day       int64  
 7   days_since_prior_order  float64
dtypes: float64(1), int64(6), object(1)
memory usage: 208.8+ MB


## *Step 4: If you find mixed-type data, fix it

There are no mixed data types within these columns. 

# Step 5

## 06. Run a check for missing values

In [33]:
# Find the missing values
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
evaluation_status              0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

**Summary of Missing Values:**
- Unnamed: 0: 0 missing values
- order_id: 0 missing values
- user_id: 0 missing values
- evaluation_status: 0 missing values
- order_number: 0 missing values
- orders_day_of_week: 0 missing values
- order_hour_of_day: 0 missing values
- days_since_prior_order: 206,209 missing values

The days_since_prior_order column is the only one with missing values. This variable might not be collected for first-time orders, which would logically result in empty entries. Alternatively, these missing values could be due to data entry errors or omissions.

# Step 6

## 07. Address the missing values

I would start with step 1 "Create a Flag for Missing Values"

In [34]:
# Step 1: Create a new variable that acts like a flag based on the missing value
df_ords['days_since_prior_order_missing'] = df_ords['days_since_prior_order'].isnull().astype(int)

I would then move on to step 2 "Impute the value with the mean or median of the column (if the variable is numeric)."

In [35]:
# Step 2: Impute the missing values with the median of the 'days_since_prior_order' column
median_value = df_ords['days_since_prior_order'].median()
df_ords['days_since_prior_order'].fillna(median_value, inplace=True)

Then verify there are no more missing values

In [36]:
# Verify that there are no more missing values
missing_values_after_imputation = df_ords.isnull().sum()
print(missing_values_after_imputation)

Unnamed: 0                        0
order_id                          0
user_id                           0
evaluation_status                 0
order_number                      0
orders_day_of_week                0
order_hour_of_day                 0
days_since_prior_order            0
days_since_prior_order_missing    0
dtype: int64


# Step 7

## 08. Run a check for duplicate values

In [37]:
# Check for duplicate values in the df_ords dataframe and store the result in a variable with 'ords' in the name
df_ords_dups = df_ords[df_ords.duplicated()]

In [38]:
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, evaluation_status, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order, days_since_prior_order_missing]
Index: []

**There are no duplicates**

However, if there were I would create a new dataframe that uses the drop_duplicates() function

## *Step 8: Address the duplicates

As mentioned in the previous step, there were no duplicates found.

# Step 9

## 09. Export the final cleaned datasets

In [39]:
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))


In [40]:
df_prods.to_csv(os.path.join(path, '02 Data','Prepared Data', 'prods_checked.csv'))
